code/__HELPERS/paths/jps.dm

```diff
@@ -0,0 +1,306 @@
+	src.caller = caller
+	src.pass_info = new(caller, access)
+	src.max_distance = max_distance
+	src.simulated_only = simulated_only
+	src.avoid = avoid
+	src.on_finish = on_finish
+	src.mintargetdist = mintargetdist
+	src.skip_first = skip_first
+	src.diagonal_handling = diagonal_handling
+	end = get_turf(goal)
+	open = new /datum/heap(/proc/HeapPathWeightCompare)
+	found_turfs = list()
+
+/datum/pathfind/jps/Destroy(force)
+	. = ..()
+	caller = null
+	end = null
+	open = null
+
+/datum/pathfind/jps/start()
+	start = start || get_turf(caller)
+	. = ..()
+	if(!.)
+		return .
+
+	if(!get_turf(end))
+		stack_trace("Invalid JPS destination")
+		return FALSE
+	if(start.z != end.z || start == end ) //no pathfinding between z levels
+		return FALSE
+	if(max_distance && (max_distance < get_dist(start, end))) //if start turf is farther than max_distance from end turf, no need to do anything
+		return FALSE
+
+	var/datum/jps_node/current_processed_node = new (start, -1, 0, end)
+	open.insert(current_processed_node)
+	found_turfs[start] = TRUE // i'm sure this is fine
+	return TRUE
+
+/datum/pathfind/jps/search_step()
+	. = ..()
+	if(!.)
+		return .
+	if(QDELETED(caller))
+		return FALSE
+
+	while(!open.is_empty() && !path)
+		var/datum/jps_node/current_processed_node = open.pop() //get the lower f_value turf in the open list
+		if(max_distance && (current_processed_node.number_tiles > max_distance))//if too many steps, don't process that path
+			continue
+
+		var/turf/current_turf = current_processed_node.tile
+		for(var/scan_direction in list(EAST, WEST, NORTH, SOUTH))
+			lateral_scan_spec(current_turf, scan_direction, current_processed_node)
+
+		for(var/scan_direction in list(NORTHEAST, SOUTHEAST, NORTHWEST, SOUTHWEST))
+			diag_scan_spec(current_turf, scan_direction, current_processed_node)
+
+		// Stable, we'll just be back later
+		if(TICK_CHECK)
+			return TRUE
+	return TRUE
+
+/datum/pathfind/jps/finished()
+	//we're done! turn our reversed path (end to start) into a path (start to end)
+	found_turfs = null
+	QDEL_NULL(open)
+
+	var/list/path = src.path || list()
+	path = reverseList(path)
+	switch(diagonal_handling)
+		if(DIAGONAL_REMOVE_CLUNKY)
+			path = remove_clunky_diagonals(path, pass_info, simulated_only, avoid)
+		if(DIAGONAL_REMOVE_ALL)
+			path = remove_diagonals(path, pass_info, simulated_only, avoid)
+	if(skip_first && length(path) > 0)
+		path.Cut(1,2)
+	hand_back(path)
+	return ..()
+
+/// Called when we've hit the goal with the node that represents the last tile, then sets the path var to that path so it can be returned by [datum/pathfind/proc/search]
+/datum/pathfind/jps/proc/unwind_path(datum/jps_node/unwind_node)
+	path = new()
+	var/turf/iter_turf = unwind_node.tile
+	path.Add(iter_turf)
+
+	while(unwind_node.previous_node)
+		var/dir_goal = get_dir(iter_turf, unwind_node.previous_node.tile)
+		for(var/i = 1 to unwind_node.jumps)
```

Comment:
we use `var/i in 1 to x` in every other instance through this PR, is there any particular reason this is `var/i = 1 to x`? i just think it'd be nice if everything were unified even if it doesn't matter in the long run. feel free to ignore if this is too nitty.

Reply:
yeah sure no, don't put yourself down this syntax sucks it's literally the same as the in pattern. I prob just copied it and didn't notice. will yeet